src/lib.rs

```diff
@@ -204,6 +203,110 @@ mod zerocopy {
     pub(crate) use crate::*;
 }
 
+/// The layout of a type which might be dynamically-sized.
+///
+/// `DstLayout` describes the layout of sized types, slice types, and "custom
+/// DSTs" - ie, those that are known by the type system to have a trailing slice
+/// (as distinguished from `dyn Trait` types - such types *might* have a
+/// trailing slice type, but the type system isn't aware of it).
+#[doc(hidden)]
+#[allow(missing_debug_implementations, missing_copy_implementations)]
+#[cfg_attr(test, derive(Debug, PartialEq, Eq))]
+pub struct DstLayout {
+    /// For sized types, `size_of::<T>()`. For DSTs, the size of the type when
+    /// the trailing slice field contains 0 elements.
+    _fixed_prefix_size: usize,
+    /// The alignment of the type.
+    _align: NonZeroUsize,
+    /// For sized types, `None`. For DSTs, the size of the element type of the
+    /// trailing slice.
+    _trailing_slice_elem_size: Option<usize>,
+}
+
+impl DstLayout {
+    /// Constructs a `DstLayout` which describes `T`.
+    ///
+    /// # Safety
+    ///
+    /// Unsafe code may assume that `DstLayout` is the correct layout for `T`.
+    const fn for_type<T>() -> DstLayout {
+        DstLayout {
+            _fixed_prefix_size: mem::size_of::<T>(),
+            _align: match NonZeroUsize::new(mem::align_of::<T>()) {
+                Some(align) => align,
+                None => panic!("core::mem::align_of should never return 0"),
+            },
+            _trailing_slice_elem_size: None,
+        }
+    }
+
+    /// Constructs a `DstLayout` which describes `[T]`.
+    ///
+    /// # Safety
+    ///
+    /// Unsafe code may assume that `DstLayout` is the correct layout for `[T]`.
+    const fn for_slice<T>() -> DstLayout {
+        DstLayout {
+            _fixed_prefix_size: 0,
+            _align: match NonZeroUsize::new(mem::align_of::<T>()) {
+                Some(align) => align,
+                None => panic!("core::mem::align_of should never return 0"),
+            },
+            _trailing_slice_elem_size: Some(mem::size_of::<T>()),
+        }
+    }
+}
+
+/// A trait which carries information about a type's layout that is used by the
+/// internals of this crate.
+///
+/// This trait is not meant for consumption by code outsie of this crate. While
+/// the normal semver stability guarantees apply with respect to which types
+/// implement this trait and which trait implementations are implied by this
+/// trait, no semver stability guarantees are made regarding its internals; they
+/// may change at any time, and code which makes use of them may break.
```

Comment:
I think this is fine for now, but I'm going to want to bikeshed this when `doc(hidden)` is removed. I think this is threading a very fine needle, and the subtleties of what's being said here aren't going to be appreciated.

Reply:
I'd be happy to take a suggested alternative wording in this PR if you have one. Also happy to punt until later.

Reply:
Let's punt until this actually gets exposed — I think I'll have better feedback then.